flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricFetcherImpl.java

```diff
@@ -148,97 +154,117 @@ private void fetchMetrics() {
+
+        return taskManagerQueryServiceGatewaysFuture.thenComposeAsync(
+                (Collection<Tuple2<ResourceID, String>> queryServiceGateways) -> {
+                    if (queryServiceGateways.isEmpty()) {
+                        return CompletableFuture.completedFuture(null);
+                    }
+                    List<CompletableFuture<Void>> queryMetricFutures = new ArrayList<>();
+                    List<String> taskManagersToRetain =
+                            queryServiceGateways.stream()
+                                    .map(
+                                            (Tuple2<ResourceID, String> tuple) -> {
+                                                CompletableFuture<Void> finishQueryMetricsFuture =
+                                                        new CompletableFuture<>();
+                                                queryServiceRetriever
+                                                        .retrieveService(tuple.f1)
+                                                        .thenAcceptAsync(
+                                                                queryServiceGateway ->
+                                                                        queryMetrics(
+                                                                                queryServiceGateway,
+                                                                                finishQueryMetricsFuture),
+                                                                executor);
+                                                queryMetricFutures.add(finishQueryMetricsFuture);
+                                                return tuple.f0.getResourceIdString();
+                                            })
+                                    .collect(Collectors.toList());
+
+                    metrics.retainTaskManagers(taskManagersToRetain);
+                    return FutureUtils.waitForAll(queryMetricFutures);
+                },
+                executor);
     }
 
     /**
      * Retrieves and queries the specified QueryServiceGateway.
      *
      * @param queryServiceAddress specifying the QueryServiceGateway
      */
-    private void retrieveAndQueryMetrics(String queryServiceAddress) {
+    private CompletableFuture<Void> retrieveAndQueryMetrics(String queryServiceAddress) {
         LOG.debug("Retrieve metric query service gateway for {}", queryServiceAddress);
 
         final CompletableFuture<MetricQueryServiceGateway> queryServiceGatewayFuture =
                 queryServiceRetriever.retrieveService(queryServiceAddress);
+        final CompletableFuture<Void> finishQueryMetricsFuture = new CompletableFuture<>();
 
         queryServiceGatewayFuture.whenCompleteAsync(
                 (MetricQueryServiceGateway queryServiceGateway, Throwable t) -> {
                     if (t != null) {
+                        finishQueryMetricsFuture.complete(null);
```

Comment:
We can make `queryMetrics` returns a `CompletableFuture`, which can be connected to `queryServiceGatewayFuture` as the next stage.